Add HeaderNames.XRequestedWith = "X-Requested-With"

ssg opened the pull request:
X-Requested-With is the standard for differentiating AJAX requests and is commonly used. Having this in HeaderNames would prevent typos related to typing this header name. I couldn't find any rationale about excluding this but there might be legitimate reasons like discouraging the use of HTTP headers or differentiating AJAX requests, etc. Please reject this if that's the case.

https://github.com/dotnet/aspnetcore/issues/19478

ssg committed:
Add HeaderNames.XRequestedWith = "X-Requested-With"

X-Requested-With is the standard for differentiating AJAX requests and is commonly used. Having this in HeaderNames would prevent typos related to typing this header name. I couldn't find any rationale about excluding this but there might be legitimate reasons like discouraging the use of HTTP headers or differentiating AJAX requests, etc. Please reject this if that's the case.

benaadams reviewed (approved):
Might want to change the explicit uses of the string in the repo to `HeaderNames.XRequestedWith` also?

e.g.
https://github.com/dotnet/aspnetcore/blob/a2568cbe1e8dd92d8a7976469100e564362f778e/src/Security/Authentication/Cookies/src/CookieAuthenticationEvents.cs#L106-L107

Though there are probably more places

ssg commented:
> Might want to change the explicit uses of the string in the repo to `HeaderNames.XRequestedWith` also?

That's a good idea but I need to set up a local build env for that as I don't want to cross dependency boundaries. Let me attend to this, thanks!

ssg committed:
Replace "X-Requested-With" references with HeaderNames.XRequestedWith

The only remaining instance is WebSocketsTransport.cs in
SignalR\clients\csharp\Http.Connections.Client which doesn't have
Microsoft.Net in its references. I didn't want to impose a new
dependency as its risky.

ssg committed:
Fix the order of using statements

ssg commented:
I replaced existing strings and built the individual projects successfully but couldn't run tests due to a problem with my dev env setup. Hopefully, CI can pick that up.

I omitted changing one file: WebSocketsTransport.cs in SignalR clients which didn't have Microsoft.Net.Http.Headers in its package references so I didn't want to add a new dependency in case it might break something.

benaadams commented:
You'll also need to add it to the ref assembly `/src/Http/Headers/ref/Microsoft.Net.Http.Headers.netcoreapp.cs`
```diff
         public static readonly string WebSocketSubProtocols;
         public static readonly string WWWAuthenticate;
         public static readonly string XFrameOptions;
+        public static readonly string XRequestedWith;
     }

```

ssg committed:
Add XRequestedWith to the ref assembly

ssg commented:
> Microsoft.Net.Http.Headers.netcoreapp.cs

Thanks! Done.

ssg commented:
Working on the test failure.

ssg commented:
The failing test (`Settings_MaxHeaderListSize_Server`) is marked as [flaky](https://github.com/dotnet/runtime/issues/860). I don't see how it's relevant to this change. It doesn't even use `HeaderNames`.

anurse commented:
@Tratcher or myself can "champion" this for API review. It seems self-explanatory and trivial :).

BrennanConroy commented:
> WebSocketsTransport.cs in SignalR clients which didn't have Microsoft.Net.Http.Headers in its package references

@Tratcher confirmed, SignalR client has very few dependencies, and this isn't one of them.

anurse commented:
Email API review yielded no concerns. Marked as API Review Approved.

anurse commented:
Thanks for the contribution @ssg!
